io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -403,9 +403,10 @@ instance MonadDelay (SimM s) where
   -- Use default in terms of MonadTimer
 
 instance MonadTimer (SimM s) where
-  data Timeout (SimM s) = Timeout !(TVar s TimeoutState) !TimeoutId
+  data Timeout (SimM s) = Timeout !(Maybe (TVar s TimeoutState)) !TimeoutId
 
-  readTimeout (Timeout var _key) = readTVar var
+  readTimeout (Timeout (Just var) _key) = readTVar var
+  readTimeout (Timeout Nothing    _key) = pure TimeoutFired
```

Comment:
Fired or cancelled?

Do we need to change the representation to use a Maybe? We could stick to the TVar but initialise it with the constant value.

Reply:
`Fired` is wrong; it breaks the `withTimeoutSerial` function; I think `TimeoutCancelled` will be good. Either will do, bug I'd  prefer to avoid a `TVar`;